src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -0,0 +1,423 @@
+                            break;
+                        case FieldSymbol field:
+                            evaluation = new BoundDagFieldEvaluation(prop.pattern.Syntax, field, input);
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+                    }
+                    decisions.Add(evaluation);
+                    var output = new BoundDagTemp(prop.pattern.Syntax, prop.symbol.GetTypeOrReturnType(), evaluation, 0);
+                    MakeDecisionsAndBindings(output, prop.pattern, decisions, bindings, ref discardedUseSiteDiagnostics);
+                }
+            }
+
+            if (recursive.VariableAccess != null)
+            {
+                // we have a "variable" declaration
+                bindings.Add((recursive.VariableAccess, input));
+            }
+        }
+
+        private static BoundDecisionDag MakeDecisionDag(ImmutableArray<PartialCaseDecision> cases)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    internal class PartialCaseDecision
+    {
+        public readonly int Index;
+        public readonly ImmutableArray<BoundDagDecision> Decisions;
+        public readonly ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings;
+        public readonly BoundExpression WhereClause;
+        public readonly LabelSymbol CaseLabel;
+        public PartialCaseDecision(
+            int Index,
+            ImmutableArray<BoundDagDecision> Decisions,
+            ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings,
+            BoundExpression WhereClause,
+            LabelSymbol CaseLabel)
+        {
+            this.Index = Index;
+            this.Decisions = Decisions;
+            this.Bindings = Bindings;
+            this.WhereClause = WhereClause;
+            this.CaseLabel = CaseLabel;
+        }
+    }
+
+    partial class BoundDagEvaluation
+    {
+        public override bool Equals(object obj) => obj is BoundDagEvaluation other && this.Equals(other);
+        public bool Equals(BoundDagEvaluation other)
+        {
+            return other != (object)null && this.Kind == other.Kind && this.Input.Equals(other.Input) && this.Symbol == other.Symbol;
+        }
+        private Symbol Symbol
+        {
+            get
+            {
+                switch (this)
+                {
+                    case BoundDagFieldEvaluation e: return e.Field;
+                    case BoundDagPropertyEvaluation e: return e.Property;
+                    case BoundDagTypeEvaluation e: return e.Type;
+                    case BoundDagDeconstructEvaluation e: return e.DeconstructMethod;
+                    default: throw ExceptionUtilities.UnexpectedValue(this.Kind);
+                }
+            }
+        }
+        public override int GetHashCode()
+        {
+            return this.Input.GetHashCode() ^ (this.Symbol?.GetHashCode() ?? 0);
```

Comment:
>this.Input.GetHashCode() ^ (this.Symbol?.GetHashCode() ?? 0); [](start = 19, length = 61)

Please use helpers to combine hash codes.

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156559981](https://github.com/dotnet/roslyn/pull/23209#discussion_r156559981) [](ancestors = 156559981)